### DISCUSSION TITLE
Retrieving the specific Invalidation of Agreements

### DISCUSSION THREAD

**Discussion by hansen-maria** [eclipse-edc/Connector#4486] [Q&A]
I am currently playing around a bit with the EDC Connector and am trying to develop my own client and connector that can also communicate with EDC Connectors, among other things.

This already works via the Management API, but when calling the protocol API (DSP interface) endpoint there are problems in places. I can start contract negotiations via the protocol API, but when I try to initiate a transfer, the connector returns the error ```409 "Cannot process TransferRequestMessage because agreement not found or not valid."```? 

Since I can retrieve an agreement with the specified ID via the Management API, I assume the second case, namely that the agreement is not valid in some way.

The agreement (retrieved via the Management API ```/v2/contractagreements/{id}```) looks like this:
```json
{
  "@context": {
    "odrl": "http://www.w3.org/ns/odrl/2/",
    "@vocab": "https://w3id.org/edc/v0.0.1/ns/",
    "edc": "https://w3id.org/edc/v0.0.1/ns/"
  },
  "@id": "086a9fca-3081-4dfd-b36d-9c36390bcbf2",
  "@type": "ContractAgreement",
  "assetId": "9120af30-d6f0-4dd6-9f30-620d18085644",
  "consumerId": "consumer",
  "contractSigningDate": 1727096138,
  "policy": {
    "@id": "50483f4f-8e20-4b23-9791-62907ae89008",
    "@type": "odrl:Agreement",
    "odrl:assignee": "consumer",
    "odrl:assigner": "provider",
    "odrl:obligation": [],
    "odrl:permission": {
      "odrl:action": {
        "@id": "odrl:use"
      }
    },
    "odrl:prohibition": [],
    "odrl:target": {
      "@id": "9120af30-d6f0-4dd6-9f30-620d18085644"
    }
  },
  "providerId": "provider"
}
```

The transfer request message as follows (POST to ```{}/transfers/request``` endpoint):
```json
{
  "@context": {
    "dspace": "https://w3id.org/dspace/v0.8/",
    "odrl": "http://www.w3.org/ns/odrl/2/"
  },
  "@type": "https://w3id.org/dspace/v0.8/TransferRequestMessage",
  "dspace:agreementId": "086a9fca-3081-4dfd-b36d-9c36390bcbf2",
  "dct:format": "example:HTTP_PUSH",
  "dspace:dataAddress": {
    "@type": "dspace:DataAddress",
    "dspace:endpointType": "https://w3id.org/idsa/v4.1/HTTP",
    "dspace:endpoint": "http://host.docker.internal:19999",
    "dspace:endpointProperties": [
      {
        "@type": "dspace:EndpointProperty",
        "dspace:name": "authorization",
        "dspace:value": "TOKEN-ABCDEFG"
      },
      {
        "@type": "dspace:EndpointProperty",
        "dspace:name": "authType",
        "dspace:value": "bearer"
      }
    ]
  },
  "dspace:callbackAddress": "http://consumer-connector:9194/protocol",
  "dspace:consumerPid": "a649b3a6-5a3d-49f0-b262-2a1769be0d41"
}
```

Is there a way to check what is not valid?

I am currently working with EDC Connectors (one as consumer, one as provider) version 0.8.1.

Many thanks in advance!

**Comment by jimmarino** [comment]
Without proper logs, stacktraces, or a precise description of the calls being made it's difficult to diagnose.

The Management API is what clients should use to interact with the connector instead of DSP messages. You should not use DSP messages unless you are writing a DSP implementation and attempting to achieve interop with the EDC. If that is the case, it's best first to gain a solid understanding of the [DSP Specification](https://github.com/eclipse-dataspace-protocol-base/DataspaceProtocol).

**Reply by hansen-maria** [reply]
Thank you very much for the quick response.

Yes, one of my goals is to implement a custom connector based on the DSP that can communicate with EDC. For this I have used the [documentation of the International Data Spaces Association](https://docs.internationaldataspaces.org/ids-knowledgebase/v/dataspace-protocol).

I have also been able to implement and use the functionalities of the Common Functionalities, Catalog and Contract Negotiations sections. It is only with the functionalities of the Transfer Process that I encounter the aforementioned problems when I want to initiate a transfer at EDC via the DSP and I receive the error that the agreement cannot be found or is invalid.

I have collected some information (debug logs of the connectors, messages that are sent and the general procedure) in a [gist](https://gist.github.com/hansen-maria/366a875ba9413766d3d7ac71ca81dda0). If more information is needed, please let me know.

**Comment by drhansbeck** [comment]
Solution is in the gist from hansen-maria: https://gist.github.com/hansen-maria/366a875ba9413766d3d7ac71ca81dda0 

_I got the problem fixed and it seemed that the problem was not the agreement or the message as such (as the error message would suggest), but rather the message header of the TransferRequestMessage. In this header, the host was not resolved correctly in my application. After correcting this, it now works and transfers can be initialized._
